### ISSUE TITLE
ONNX serialization/deserialization

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#57]
Would be useful to export/import models to onnx

https://github.com/onnx/onnx

**Comment by coreylowman** [comment]
Whoa this is so useful thanks!

**Comment by M1ngXU** [comment]
is this still planned to be added?

**Comment by coreylowman** [comment]
Its not highest priority, but yes

**Comment by M1ngXU** [comment]
if i'd like to implement this, would this be some kind of own model struct with some backend (like ort/wonnx) used?

**Comment by M1ngXU** [comment]
or parse the model to the specified type and panic if it doesn't work?

**Comment by coreylowman** [comment]
It would be saving an insantiated model to an `.onnx` file, or loading from one.

THis would probably be some core traits like:
```rust
trait SaveToONNX { }
trait LoadFromONNX {}
```

and then we'd have to implement the above traits for each module we have (can all be in one onnx.rs file instead of spread out).

One complication is figuring out how we are going to handle the different onnx opsets. Maybe to start we just assume the latest opset version and stick with that.

Looks like we may want to depend on tract's onnx crate for helping create onnx models https://docs.rs/tract-onnx/0.19.7/tract_onnx/.

I'd say first goal can be saving/loading a Linear layer to a `.onnx` file to flesh out the traits, and then we can add impls for the other layers from there.

**Comment by M1ngXU** [comment]
oof, seems to be complicated. maybe i'll stick with a (worse) but "pure" onnx crate lol

**Comment by coreylowman** [comment]
Hah yeah onnx is quite a beast

**Comment by ianpurton** [comment]
Burn is starting to add ONNX import. https://github.com/burn-rs/burn/tree/main/burn-import/src/onnx

Looks like they're using https://github.com/stepancheg/rust-protobuf/ to generate rust code from the onnx.proto file then they generate source code that corresponds to what you would write in burn.

They say the won't implement ONNX export.

I would actually start with export as it gives a nice feature of being able to build models in rust but deploy using ONNX runtime.

So what would be a good strategy for implementing export?

Is it about implementing the https://github.com/coreylowman/dfdx/blob/main/src/tensor/storage_traits.rs for ONNX?

**Comment by M1ngXU** [comment]
> Is it about implementing the https://github.com/coreylowman/dfdx/blob/main/src/tensor/storage_traits.rs for ONNX?

i don't think so, this would be required if we use onnx as some kind of device (i think). not sure how the onnx format works, but doing something with the tensorcollection to convert a model to onnx makes sense. not sure about custom models though.



> Burn is starting to add ONNX import. https://github.com/burn-rs/burn/tree/main/burn-import/src/onnx
> 
> Looks like they're using https://github.com/stepancheg/rust-protobuf/ to generate rust code from the onnx.proto file then they generate source code that corresponds to what you would write in burn.

burn has always been weird lol, saying that dynamic dimensions is better than const ones etc.

**Comment by ianpurton** [comment]
> not sure how the onnx format works

It's a protobuf definition https://github.com/onnx/onnx/blob/main/onnx/onnx.proto

The way I figured it out is to look at the wonnx examples https://github.com/webonnx/wonnx/blob/master/wonnx/examples/simple_graph.rs. 

So the rust below is converted directly to protobuf by wonnx and will then run on onnx runtimes.

```rust
// ONNX INPUTS
  let shape = vec![1, c, n as i64, n as i64];
  let kernel_n = 3;
  let m = 1;
  let data_w: Vec<f32> = (0..m * c * kernel_n * kernel_n).map(|_| 1.0f32).collect();
  let model = model(graph(
      vec![tensor("X", &shape)],
      vec![tensor("Y", &[1, m, 3, 3])],
      vec![],
      vec![initializer("W", data_w, vec![m, c, 3, 3])],
      vec![node(
          vec!["X", "W"],
          vec!["Y"],
          "conv",
          "Conv",
          vec![attribute("kernel_shape", vec![3, 3])],
      )],
  ));
```

So would there be a way where we could somehow get the model (compute graph) from dfdx?

**Comment by M1ngXU** [comment]
this looks like it can be directly converted using `TensorCollection`s. Though I don’t know the details about how to use that. Maybe you can have a look at `NumParams` (or so) how it’s done there?

**Comment by ianpurton** [comment]
Thanks, for reference it's here https://github.com/coreylowman/dfdx/blob/main/src/nn/num_params.rs

Looks like a visitor pattern, so if it does visit all of the model (recursively?) then there might be a mapping to something like the above.

**Comment by M1ngXU** [comment]
yes, but not sure how dfdx modules/tensors map to onnx stuff 😅. but his should be the correct way 👍

**Comment by coreylowman** [comment]
Yeah it's a visitor pattern, however it's a tensor visitor, so the current APIs don't know what type of module the tensor is contained in.

In my head whatever onnx export traits would have to be implemented separately for each module type. So we'd have something like:

```rust
impl<const I: usize, const O: usize, E: Dtype, D: Device<E>> OnnxExport for Linear<I, O, E, D> {
    fn export(...) { ... }
}

impl<F: OnnxExport> OnnxExport for Residual<F> {
    fn export(...) { ... }
}

... // and many more
```
